src/bindings/python/flux/job.py

```diff
@@ -85,3 +96,385 @@ def wait(flux_handle, jobid=lib.FLUX_JOBID_ANY):
+            elif range_dict[key] < 1:
+                raise ValueError("{} must be > 0".format(key))
+        valid_operator_values = ["+", "*", "^"]
+        if (
+            "operator" in range_dict
+            and range_dict["operator"] not in valid_operator_values
+        ):
+            raise ValueError("operator must be one of {}".format(valid_operator_values))
+
+    @classmethod
+    def _validate_resource(cls, res):
+        if not isinstance(res, collectionsAbc.Mapping):
+            raise TypeError("resource must be a mapping")
+
+        # validate the 'type' key
+        if "type" not in res:
+            raise ValueError("type is a required key for resources")
+        if not isinstance(res["type"], six.string_types):
+            raise TypeError("type must be a string")
+
+        # validate the 'count' key
+        if "count" not in res:
+            raise ValueError("count is a required key for resources")
+        count = res["count"]
+        if isinstance(count, collectionsAbc.Mapping):
+            cls._validate_complex_range(count)
+        elif not isinstance(count, six.integer_types):
+            raise TypeError("count must be an int or mapping")
+        elif count < 1:
+            raise ValueError("count must be > 0")
+
+        # validate the string keys
+        for key in ["id", "unit", "label"]:
+            if key in res and not isinstance(res[key], six.string_types):
+                raise TypeError("{} must be a string".format(key))
+
+        # validate the 'exclusive' key
+        if "exclusive" in res:
+            if res["exclusive"] not in [True, False]:
+                raise TypeError("exclusive must be a boolean")
+
+        # validate that slots have a 'label'
+        if res["type"] == "slot" and "label" not in res:
+            raise ValueError("slots must have labels")
+
+    @staticmethod
+    def _validate_task(task):
+        if not isinstance(task, collectionsAbc.Mapping):
+            raise TypeError("task must be a mapping")
+
+        _validate_keys(["command", "slot", "count"], task.keys(), allow_additional=True)
+
+        if not isinstance(task["count"], collectionsAbc.Mapping):
+            raise TypeError("count must be a mapping")
+
+        if not isinstance(task["slot"], six.string_types):
+            raise TypeError("slot must be a string")
+
+        if "attributes" in task and not isinstance(
+            task["attributes"], collectionsAbc.Mapping
+        ):
+            raise TypeError("count must be a mapping")
+
+        command = task["command"]
+        if not (
+            (  # sequence of strings - N.B. also true for just a plain string
+                isinstance(command, collectionsAbc.Sequence)
+                and all(isinstance(x, six.string_types) for x in command)
+            )
+        ) or isinstance(command, six.string_types):
+            raise TypeError("command must be a list of strings")
+
+    @staticmethod
+    def _validate_attributes(attributes):
```

Comment:
Do we only want to allow these attributes? I kinda thought this was a general set, but maybe misunderstood.

Reply:
Based on how RFC14 is written, I believe that these are the only two valid values, but users can put whatever they want within them.

From RFC14:
> The value of the attributes key SHALL be a dictionary of dictionaries. The attributes dictionary MAY contain one or both of the following keys which, if present, must have values. Values MAY have any valid YAML type.

Reply:
Interesting, ok.  We'll probably want to loosen that for non-v1 but leave this as is since it matches.  Thanks!